protocols/rendezvous/src/substream_handler.rs

```diff
@@ -0,0 +1,463 @@
+    Done,
+}
+
+impl<TState, TEvent> Next<TState, TEvent> {
+    pub fn map_state<TNextState>(
+        self,
+        map: impl FnOnce(TState) -> TNextState,
+    ) -> Next<TNextState, TEvent> {
+        match self {
+            Next::EmitEvent { event, next_state } => Next::EmitEvent {
+                event,
+                next_state: map(next_state),
+            },
+            Next::Pending { next_state } => Next::Pending {
+                next_state: map(next_state),
+            },
+            Next::Continue { next_state } => Next::Pending {
+                next_state: map(next_state),
+            },
+            Next::Done => Next::Done,
+        }
+    }
+}
+
+#[derive(Debug, Hash, Eq, PartialEq, Clone, Copy)]
+pub struct InboundSubstreamId(u64);
+
+impl InboundSubstreamId {
+    fn fetch_and_increment(&mut self) -> Self {
+        let next_id = *self;
+        self.0 += 1;
+
+        next_id
+    }
+}
+
+impl fmt::Display for InboundSubstreamId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[derive(Debug, Hash, Eq, PartialEq, Clone, Copy)]
+pub struct OutboundSubstreamId(u64);
+
+impl OutboundSubstreamId {
+    fn fetch_and_increment(&mut self) -> Self {
+        let next_id = *self;
+        self.0 += 1;
+
+        next_id
+    }
+}
+
+impl fmt::Display for OutboundSubstreamId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+type ProtocolUpgradeFn =
+    Box<dyn Fn(NegotiatedSubstream, Endpoint) -> Ready<Result<NegotiatedSubstream, Void>> + Send>;
+type Protocol = FromFnUpgrade<&'static [u8], ProtocolUpgradeFn>;
+
+/// An implementation of [`ProtocolsHandler`] that delegates to individual [`SubstreamHandler`]s.
+pub struct SubstreamProtocolsHandler<TInboundSubstream, TOutboundSubstream, TOutboundOpenInfo> {
+    inbound_substreams: HashMap<InboundSubstreamId, TInboundSubstream>,
+    outbound_substreams: HashMap<OutboundSubstreamId, TOutboundSubstream>,
+    protocol: &'static [u8],
+    next_inbound_substream_id: InboundSubstreamId,
+    next_outbound_substream_id: OutboundSubstreamId,
+
+    new_substreams: VecDeque<TOutboundOpenInfo>,
+
+    initial_keep_alive_deadline: Instant,
+}
+
+impl<TInboundSubstream, TOutboundSubstream, TOutboundOpenInfo>
+    SubstreamProtocolsHandler<TInboundSubstream, TOutboundSubstream, TOutboundOpenInfo>
+{
+    pub fn new(protocol: &'static [u8], initial_keep_alive: Duration) -> Self {
+        Self {
+            inbound_substreams: Default::default(),
+            outbound_substreams: Default::default(),
+            protocol,
+            next_inbound_substream_id: InboundSubstreamId(0),
+            next_outbound_substream_id: OutboundSubstreamId(0),
+            new_substreams: Default::default(),
+            initial_keep_alive_deadline: Instant::now() + initial_keep_alive,
+        }
+    }
+}
+
+fn poll_substreams<TId, TSubstream, TError, TOutEvent>(
```

Comment:
It could be but only as an associated function because it is being called with both `&mut self.inbound_substreams` and `&mut self.outbound_substreams`. Hence, I didn't see an advantage of putting it within an `impl SubstreamProtocolsHandler`.

Reply:
Might be worth a comment :)